src/schema/posts.ts

```diff
@@ -1327,6 +1366,76 @@ export const resolvers: IResolvers<any, Context> = {
       }
       return { _: true };
     },
+    downvote: async (
+      source,
+      { id }: { id: string },
+      ctx: Context,
+    ): Promise<GQLEmptyResponse> => {
+      try {
+        const post = await ctx.con.getRepository(Post).findOneByOrFail({ id });
+        await ensureSourcePermissions(ctx, post.sourceId);
+        await ctx.con.transaction(async (entityManager) => {
+          await entityManager.getRepository(Downvote).insert({
+            postId: id,
+            userId: ctx.userId,
+          });
+
+          await saveHiddenPost(entityManager.connection, {
+            userId: ctx.userId,
+            postId: id,
+          });
+
+          const upvote = entityManager.getRepository(Upvote).findOneBy({
+            postId: id,
+            userId: ctx.userId,
+          });
+
+          if (upvote) {
+            await entityManager.getRepository(Upvote).delete({
+              postId: id,
+              userId: ctx.userId,
+            });
+          }
```

Comment:
We can just delete the object without checking if it exists to reduce number of queries

Reply:
@idoshamun there is a simular check if `cancelUpvote`, we can then remove it from there also?

Reply:
Basically yes